[19.09] Install manager unit test and python 3 hg util exception handling fixes

mvdbeek opened the pull request:
On the install manager side these are only cosmetic fixes in normal operation, but I think this has the potential to be buggy when errors occur on python 3. Same for the hg_util fixes.
The unit test is more of a poor excuse, but at least it exercises installation and update (with the actual cloning and TS interaction mocked out and pretending there is no tool to load).

mvdbeek committed:
Use kwargs for long TS methods

Was checking this during the review anyway, might as well add it in the
code.

mvdbeek committed:
Simplify create_or_update_tool_shed_repository

mvdbeek committed:
Drop unnecessary casting and improve exception logging

mvdbeek committed:
Add basic test for create_or_update_tool_shed_repository

mvdbeek committed:
Fix exception handling in hg_util on python 3

mvdbeek committed:
Add basic unit test for install repository manager

nsoranzo commented on lib/tool_shed/galaxy_install/repository_dependencies/repository_dependency_manager.py:
```diff
@@ -309,8 +309,8 @@ def get_repository_dependencies_for_installed_tool_shed_repository(self, app, re
-        except Exception as e:
-            log.error("The URL\n%s\nraised the exception:\n%s\n", build_url(tool_shed_url, pathspec=pathspec, params=params), str(e))
+        except Exception:
+            log.exception("The URL\n%s\nraised the exception:\n%s\n", build_url(tool_shed_url, pathspec=pathspec, params=params))
```
```suggestion
            log.exception("Error while trying to get URL: %s", build_url(tool_shed_url, pathspec=pathspec, params=params))
```

nsoranzo commented on lib/tool_shed/galaxy_install/install_manager.py:
```diff
@@ -453,17 +453,17 @@ def get_repository_components_for_installation(self, encoded_tsr_id, encoded_tsr
-            message += "%s for revision %s of repository %s owned by %s: %s" % \
-                (str(tool_shed_url), str(changeset_revision), str(name), str(owner), str(e))
-            log.warning(message)
+            message += "%s for revision %s of repository %s owned by %s:" % \
```
```suggestion
            message += "%s for revision %s of repository %s owned by %s" % \
```

mvdbeek committed:
Improve phrasing of exception messages

Co-Authored-By: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek committed:
Fix kwargs

mvdbeek committed:
Cast `ctx_rev` to `str` before passing it to subprocess command

In some places `ctx_rev` is an `int`.
Fixes the toolshed tests. Uncovered by removing the string casting in
commit 732874badba3a60599fbf19a93af3c8aa85394d5 .

nsoranzo commented on lib/tool_shed/util/hg_util.py:
```diff
@@ -395,21 +396,21 @@ def init_repository(repo_path):
 def changeset2rev(repo_path, changeset_revision):
     """
-    Return the revision number corresponding to a specified changeset revision.
+    Return the numeric revision number (as a string) corresponding to a specified changeset revision.
```
Numeric number doesn't read well
```suggestion
    Return the revision number (as a string) corresponding to a specified changeset revision.
```

nsoranzo commented on lib/tool_shed/util/repository_util.py:
```diff
@@ -89,6 +89,7 @@ def create_or_update_tool_shed_repository(app, name, description, installed_chan
     If a record defined by the received tool shed, repository name and owner does not exist, create
     a new record with the received information.
     """
+    metadata_dict = metadata_dict if metadata_dict is not None else {}
```
```suggestion
    metadata_dict = metadata_dict or {}
```

nsoranzo commented on lib/tool_shed/util/hg_util.py:
```diff
@@ -395,21 +396,21 @@ def init_repository(repo_path):
+            error_message += "\nOutput was:\n%s" % unicodify(e.output)
         raise Exception(error_message)
-    return int(rev.strip())
+    return rev.strip()
```
What was the issue with this? I think I prefer the way it was, with casting to `str` by the caller when needed, especially since:
- `subprocess.check_output()` return a bytestring
- casting to int here finds any issue as early as possible.

mvdbeek replied:
There's only one place in the code where we need `ctx_rev` to be numeric (RepositoryMetadata.numeric_revions), while everywhere else it should be a string.
I agree though that it's nice to check this is an integer,
so how about `return str(int(rev.strip())))` ?

nsoranzo replied:
`changeset2rev` is called only in 3 places, so it's not a big problem to cast in the caller if needed. Anyway, I think I've fixed the toolshed tests by casting to `str` where needed in an updated commit, let me know if you dislike it!

mvdbeek replied:
I would prefer it if ctx_rev was always a string, so I did that in my last commit. Does that seem like a reasonable compromise ?

nsoranzo committed:
Remove unused code

nsoranzo committed:
Simplify code

mvdbeek committed:
Cast ctx_rev coming from changset2rev to string
